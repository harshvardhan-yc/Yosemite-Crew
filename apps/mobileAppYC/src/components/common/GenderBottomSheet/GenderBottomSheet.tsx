import React, {forwardRef, useImperativeHandle, useRef} from 'react';
import {GenericSelectBottomSheet, type SelectItem} from '../GenericSelectBottomSheet/GenericSelectBottomSheet';
import type {CompanionGender} from '@/features/companion/types';

export interface GenderBottomSheetRef {
  open: () => void;
  close: () => void;
}

export const GenderBottomSheet = forwardRef<
  GenderBottomSheetRef,
  {
    selected?: CompanionGender | null;
    selectedGender?: CompanionGender | null;
    onSave: (g: CompanionGender) => void;
  }
>(({selected, selectedGender, onSave}, ref) => {
  const bottomSheetRef = useRef<any>(null);

  const genderItems: SelectItem[] = [
    {id: 'male', label: 'Male'},
    {id: 'female', label: 'Female'},
  ];

  const effectiveSelection = selected ?? selectedGender ?? null;
  const selectedItem = effectiveSelection
    ? genderItems.find(item => item.id === effectiveSelection) || null
    : null;

  useImperativeHandle(ref, () => ({
    open: () => {
      bottomSheetRef.current?.open();
    },
    close: () => {
      bottomSheetRef.current?.close();
    },
  }));

  const handleSave = (item: SelectItem | null) => {
    if (item) {
      onSave(item.id as CompanionGender);
    }
  };

  return (
    <GenericSelectBottomSheet
      ref={bottomSheetRef}
      title="Select Gender"
      items={genderItems}
      selectedItem={selectedItem}
      onSave={handleSave}
      hasSearch={false}
      emptyMessage="No gender options available"
      mode="select"
      snapPoints={['30%', '35%']}
      maxListHeight={300}
    />
  );
});

GenderBottomSheet.displayName = 'GenderBottomSheet';

export default GenderBottomSheet;
