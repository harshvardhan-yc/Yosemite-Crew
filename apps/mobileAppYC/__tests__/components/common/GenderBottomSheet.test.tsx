import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {GenderBottomSheet} from '@/shared/components/common/GenderBottomSheet/GenderBottomSheet';

jest.mock('@/shared/components/common/BottomSheet/BottomSheet', () => {
  const ReactModule = require('react');
  return {
    __esModule: true,
    default: ReactModule.forwardRef(({children, title}: any, ref: any) => {
      const {View, Text} = require('react-native');
      ReactModule.useImperativeHandle(ref, () => ({
        open: jest.fn(),
        close: jest.fn(),
      }));
      return ReactModule.createElement(
        View,
        {testID: 'bottom-sheet'},
        ReactModule.createElement(Text, {testID: 'sheet-title'}, title),
        children
      );
    }),
  };
});

jest.mock('@/shared/components/common/LiquidGlassButton/LiquidGlassButton', () => {
  const ReactModule = require('react');
  return {
    __esModule: true,
    default: ({title, onPress}: any) => {
      const {TouchableOpacity, Text} = require('react-native');
      return ReactModule.createElement(
        TouchableOpacity,
        {onPress, testID: `button-${title}`},
        ReactModule.createElement(Text, null, title)
      );
    },
  };
});

jest.mock('@/hooks', () => ({
  useTheme: () => ({
    theme: {
      colors: {
        primary: '#007AFF',
        secondary: '#333',
        background: '#FFF',
        cardBackground: '#FFF',
        border: '#E0E0E0',
        textSecondary: '#666',
        text: '#000',
      },
      spacing: {2: 8, 3: 12, 4: 16, 6: 24, 8: 32},
      typography: {
        bodyMedium: {fontSize: 14},
        titleMedium: {fontSize: 16},
        h3: {fontSize: 26, lineHeight: 31.2, fontWeight: '500'},
      },
      borderRadius: {md: 8, full: 9999},
    },
  }),
}));

jest.mock('@/assets/images', () => ({
  Images: {checkIcon: 1},
}));

describe('GenderBottomSheet', () => {
  const mockOnSave = jest.fn();
  const mockRef = React.createRef<any>();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders without crashing', () => {
    const {getByTestId} = render(
      <GenderBottomSheet ref={mockRef} selectedGender={null} onSave={mockOnSave} />
    );
    expect(getByTestId('bottom-sheet')).toBeTruthy();
  });

  it('renders title', () => {
    const {getByTestId} = render(
      <GenderBottomSheet ref={mockRef} selectedGender={null} onSave={mockOnSave} />
    );
    expect(getByTestId('sheet-title')).toBeTruthy();
  });

  it('renders gender options', () => {
    const {getByText} = render(
      <GenderBottomSheet ref={mockRef} selectedGender={null} onSave={mockOnSave} />
    );
    expect(getByText('Male')).toBeTruthy();
    expect(getByText('Female')).toBeTruthy();
  });

  it('calls onSave with male when Male selected', () => {
    const {getByText} = render(
      <GenderBottomSheet ref={mockRef} selectedGender={null} onSave={mockOnSave} />
    );
    fireEvent.press(getByText('Male'));
    expect(mockOnSave).toHaveBeenCalledWith('male');
  });

  it('calls onSave with female when Female selected', () => {
    const {getByText} = render(
      <GenderBottomSheet ref={mockRef} selectedGender={null} onSave={mockOnSave} />
    );
    fireEvent.press(getByText('Female'));
    expect(mockOnSave).toHaveBeenCalledWith('female');
  });

  it('shows selected state for male', () => {
    const {getByText} = render(
      <GenderBottomSheet ref={mockRef} selectedGender="male" onSave={mockOnSave} />
    );
    expect(getByText('Male')).toBeTruthy();
  });

  it('shows selected state for female', () => {
    const {getByText} = render(
      <GenderBottomSheet ref={mockRef} selectedGender="female" onSave={mockOnSave} />
    );
    expect(getByText('Female')).toBeTruthy();
  });

  it('exposes open method via ref', () => {
    render(
      <GenderBottomSheet ref={mockRef} selectedGender={null} onSave={mockOnSave} />
    );
    expect(mockRef.current?.open).toBeDefined();
  });

  it('exposes close method via ref', () => {
    render(
      <GenderBottomSheet ref={mockRef} selectedGender={null} onSave={mockOnSave} />
    );
    expect(mockRef.current?.close).toBeDefined();
  });

  it('matches snapshot', () => {
    const {toJSON} = render(
      <GenderBottomSheet ref={mockRef} selectedGender={null} onSave={mockOnSave} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
