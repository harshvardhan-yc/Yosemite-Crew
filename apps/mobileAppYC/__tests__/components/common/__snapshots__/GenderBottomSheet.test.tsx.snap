// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenderBottomSheet matches snapshot 1`] = `
<View
  testID="bottom-sheet"
>
  <Text
    testID="sheet-title"
  />
  <View
    style={
      {
        "backgroundColor": "#FFF",
        "flex": 1,
        "paddingHorizontal": undefined,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingVertical": 16,
          "position": "relative",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "paddingRight": 32,
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontSize": 26,
              "fontWeight": "500",
              "lineHeight": 33.800000000000004,
              "textAlign": "center",
            }
          }
        >
          Select Gender
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 8,
          "maxHeight": 300,
        }
      }
    >
      <RCTScrollView
        ListEmptyComponent={[Function]}
        contentContainerStyle={
          {
            "paddingBottom": 8,
          }
        }
        data={
          [
            {
              "id": "male",
              "label": "Male",
            },
            {
              "id": "female",
              "label": "Female",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        nestedScrollEnabled={true}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        showsVerticalScrollIndicator={true}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#E0E0E0",
                  "borderBottomWidth": 1,
                  "borderRadius": undefined,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginBottom": undefined,
                  "opacity": 1,
                  "paddingHorizontal": 12,
                  "paddingVertical": 14,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#000",
                    "flex": 1,
                  }
                }
              >
                Male
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#E0E0E0",
                  "borderBottomWidth": 1,
                  "borderRadius": undefined,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginBottom": undefined,
                  "opacity": 1,
                  "paddingHorizontal": 12,
                  "paddingVertical": 14,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#000",
                    "flex": 1,
                  }
                }
              >
                Female
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
